datajoint/external.py

```diff
@@ -45,44 +45,82 @@ def __init__(self, connection, store=None, database=None):
         timestamp=CURRENT_TIMESTAMP  :timestamp   # automatic timestamp
         """
 
     @property
     def table_name(self):
         return '{external_table_root}_{store}'.format(external_table_root=EXTERNAL_TABLE_ROOT, store=self.store)
 
+    @property
+    def s3(self):
+        if self._s3 is None:
+            self._s3 = s3.Folder(**self.spec)
+        return self._s3
+
     def put(self, blob):
         """
-        put an object in external store
+        put a binary string in external store
         """
-        blob_hash = uuid_from_buffer(blob)
-        if self.spec['protocol'] == 'file':
-            folder = os.path.join(self.spec['location'], self.database, *subfold(blob_hash.hex, self.spec['subfolding']))
-            full_path = os.path.join(folder, blob_hash.hex)
-            if not os.path.isfile(full_path):
-                try:
-                    safe_write(full_path, blob)
-                except FileNotFoundError:
-                    os.makedirs(folder)
-                    safe_write(full_path, blob)
-        elif self.spec['protocol'] == 's3':
-            folder = '/'.join(subfold(blob_hash.hex, self.spec['subfolding']))
-            s3.Folder(database=self.database, **self.spec).put('/'.join((folder, blob_hash.hex)), blob)
-
+        uuid = uuid_from_buffer(blob)
+        if self.spec['protocol'] == 's3':
+            self.s3.put('/'.join((self.database, '/'.join(subfold(uuid.hex, self.spec['subfolding'])), uuid.hex)), blob)
+        else:
+            remote_file = os.path.join(os.path.join(
+                self.spec['location'], self.database, *subfold(uuid.hex, self.spec['subfolding'])), uuid.hex)
+            safe_write(remote_file, blob)
         # insert tracking info
         self.connection.query(
-            "INSERT INTO {tab} (hash, size) VALUES (%s, {size}) "
-            "ON DUPLICATE KEY UPDATE timestamp=CURRENT_TIMESTAMP".format(
-                tab=self.full_table_name, size=len(blob)), args=(blob_hash.bytes,))
-        return blob_hash
+            "INSERT INTO {tab} (hash, size) VALUES (%s, {size}) ON DUPLICATE KEY "
+            "UPDATE timestamp=CURRENT_TIMESTAMP".format(
+                tab=self.full_table_name, size=len(blob)), args=(uuid.bytes,))
+        return uuid
+
+    def fput(self, local_filepath):
+        """
+        Raise exception if an external entry already exists with a different contents checksum.
+        Otherwise, copy (with overwrite) file to remote and
+        If an external entry exists with the same checksum, then no copying should occur
+        """
+        local_folder = os.path.dirname(local_filepath)
+        stage_folder = os.path.join(os.path.abspath(self.spec['stage']), '')
+        if not local_folder.startswith(stage_folder):
+            raise DataJointError('The path {path} is not in stage {stage}'.format(
+                path=local_folder, stage=stage_folder))
+        relative_filepath = local_filepath[len(stage_folder):]
```

Comment:
Can we avoid string operations on `local_filepath`? `fput` assumes this is an absolute path. What about user-specified relative paths, symbolic links? We can add tests for those cases.

Reply:
We always convert to absolute path before any operations.

Reply:
Per discussion, agreed that `os.path.relpath` implementation here would be an improvement.